src/ILCompiler.ReadyToRun/src/Compiler/ReadyToRunMetadataFieldLayoutAlgorithm.cs

```diff
@@ -4,44 +4,257 @@
-            layout.ThreadNonGcStatics.Size = LayoutInt.Zero;
-            layout.ThreadGcStatics.Size = LayoutInt.Zero;
+            /// <summary>
+            /// <a href="https://github.com/dotnet/coreclr/blob/master/src/vm/class.h#L64">MAX_LOG2_PRIMITIVE_FIELD_SIZE</a>
+            /// </summary>
+            private const int MaxLog2PrimitiveFieldSize = 3;
+
+            /// <summary>
+            /// <a href="https://github.com/dotnet/coreclr/blob/master/src/vm/class.h#L66">MAX_PRIMITIVE_FIELD_SIZE</a>
+            /// </summary>
+            private const int MaxPrimitiveFieldSize = 1 << MaxLog2PrimitiveFieldSize;
+
+            /// <summary>
+            /// CoreCLR DomainLocalModule::OffsetOfDataBlob() / sizeof(void *)
+            /// </summary>
+            private const int DomainLocalModuleDataBlobOffsetAsIntPtrCount = 6;
+
+            /// <summary>
+            /// CoreCLR ThreadLocalModule::OffsetOfDataBlob() / sizeof(void *)
+            /// </summary>
+            private const int ThreadLocalModuleDataBlobOffsetAsIntPtrCount = 3;
+
+            /// <summary>
+            /// TargetDetails is used to query pointer size for the target architecture.
+            /// </summary>
+            private TargetDetails _target;
+
+            public ModuleFieldLayoutMap(TargetDetails target)
+            {
+                _target = target;
+            }
+
+            protected override bool CompareKeyToValue(EcmaModule key, ModuleFieldLayout value)
+            {
+                return key == value.Module;
+            }
+
+            protected override bool CompareValueToValue(ModuleFieldLayout value1, ModuleFieldLayout value2)
+            {
+                return value1.Module == value2.Module;
+            }
+
+            protected override ModuleFieldLayout CreateValueFromKey(EcmaModule module)
+            {
+                int typeCountInModule = module.MetadataReader.GetTableRowCount(TableIndex.TypeDef);
+
+                // 0 corresponds to "normal" statics, 1 to thread-local statics
+                LayoutInt[] gcStatics = new LayoutInt[2]
+                {
+                    LayoutInt.Zero,
+                    LayoutInt.Zero
+                };
+                LayoutInt[] nonGcStatics = new LayoutInt[2]
+                {
+                    new LayoutInt(DomainLocalModuleDataBlobOffsetAsIntPtrCount * _target.PointerSize + typeCountInModule),
+                    new LayoutInt(ThreadLocalModuleDataBlobOffsetAsIntPtrCount * _target.PointerSize + typeCountInModule),
+                };
+                Dictionary<DefType, FieldAndOffset[]> typeToFieldMap = new Dictionary<DefType, FieldAndOffset[]>();
+
+                for (int typeIndex = 1; typeIndex <= typeCountInModule; typeIndex++)
+                {
+                    List<FieldAndOffset> fieldsForType = null;
+                    DefType defType = (DefType)module.GetObject(MetadataTokens.TypeDefinitionHandle(typeIndex));
```

Comment:
I'm wondering whether we should do this using the metadata APIs, same as what CoreCLR does. Avoiding the type system lets CoreCLR not fail the layout if there's a bogus type in there somewhere.

Having the type system involved means that this is going throw if e.g. there's a field with a type that doesn't resolve. From our .NET Native experience we know people like to use mismatched assemblies they gathered from random places on their filesystem if they work well enough for their scenario (they never execute the mismatched code).

Reply:
Hopefully fixed.